### PR TITLE
GUAC-240: Implement adaptive JPEG encoding

### PR DESCRIPTION
**Pull request by mike-jumper**
This change merges the changes from #66 while also simplifying its logic. Overall, the intent of these changes is to send updates as JPEG when doing so is appropriate because of heuristics which track:

1. The local framerate of parts of the Guacamole display
2. The complexity of updates (estimating their likelihood to compress well with PNG).

The heuristics are currently written to send updates as JPEG if they occur within a region updating at roughly 3 FPS and do not contain enough repeating runs of identical pixels to compress better with PNG.

### CODE REVIEW

File: src/common/guac_surface.c [hunk 1]
```diff
@@ -1054,13 +1347,23 @@ void guac_common_surface_flush(guac_common_surface* surface) {
...
             if ((combined > 1 || i < original_queue_length)
-                    && surface->png_queue_length < GUAC_COMMON_SURFACE_QUEUE_SIZE)
+                    && surface->bitmap_queue_length < GUAC_COMMON_SURFACE_QUEUE_SIZE)
                 __guac_common_surface_flush_to_queue(surface);
 
-            /* Flush as PNG otherwise */
-            else {
-                if (surface->dirty) flushed++;
-                __guac_common_surface_flush_to_png(surface);
+            /* Flush as bitmap otherwise */
```
Comment (jmuehlner) [inline src/common/guac_surface.c]: bitmap?

File: src/common/guac_surface.c [hunk 2]
```diff
@@ -665,19 +909,20 @@ guac_common_surface* guac_common_surface_alloc(guac_client* client,
         guac_socket* socket, const guac_layer* layer, int w, int h) {
 
     /* Init surface */
-    guac_common_surface* surface = malloc(sizeof(guac_common_surface));
+    guac_common_surface* surface = calloc(1, sizeof(guac_common_surface));
```
Comment (jmuehlner) [inline src/common/guac_surface.c]: Why use calloc here instead of malloc?

File: src/common/guac_surface.c (hunk 1 above)
Comment (mike-jumper) [inline src/common/guac_surface.c]: Yes. The flush is no longer specific to any one format of image, hence the rename from PNG-specific comments and structure to, simply, "bitmap".

File: src/common/guac_surface.c (hunk 2 above)
Comment (mike-jumper) [inline src/common/guac_surface.c]: For the automatic zeroing of memory. From the `calloc()` section of the `malloc()` man page:

> ### NAME
> 
> malloc, free, calloc, realloc - allocate and free dynamic memory
> 
> ### SYNOPSIS
>     #include <stdlib.h>
> 
>     void *malloc(size_t size);
>     void free(void *ptr);
>     void *calloc(size_t nmemb, size_t size);
>     void *realloc(void *ptr, size_t size);

> ### DESCRIPTION
> 
> ...
> 
> The `calloc()`  function allocates memory for an array of `nmemb` elements of size bytes each and returns a pointer to the allocated memory. **The memory  is  set  to zero.**  If `nmemb` or `size` is 0, then `calloc()` returns either `NULL`, or a unique pointer value that can later  be  successfully passed to `free()`.
> 
> ...
> 

Emphasis added. No such guarantee in normal `malloc()`. This is often faster than `malloc()` + `memset()`, as the kernel keeps a zeroed pages of memory available, and `calloc()` can exploit this. More on that: http://stackoverflow.com/questions/2688466/why-mallocmemset-is-slower-than-calloc

Comment (mike-jumper) [inline src/common/guac_surface.c]: FYI, also from that man page:

> ### CONFORMING TO
> 
> C89, C99.
> 

Our code is written conforming to C99 and POSIX, so we're good here.

**Pull request merged**